main/src/ca/uwaterloo/flix/language/phase/jvm/GenFunAndClosureClasses.scala

```diff
@@ -90,68 +124,110 @@ object GenFunAndClosureClasses {
     *
     *     EnterLabel:
     *
-    *     int ? = this.clo0;
-    *     byte ? = this.clo1;
     *     Object ? = this.arg0;
     *     int ? = this.arg1;
     *
     *     // body code ...
     *   }
     *
-    *   public final (Def$example|Clo$example$152) copy {
-    *     (Def$example|Clo$example$152) x = new (Def$example|Clo$example$152)();
+    *   public final Def$example copy {
+    *     Def$example x = new Def$example();
     *     x.arg0 = this.arg0;
     *     x.arg1 = this.arg1
-    *     x.clo0 = this.clo0;
-    *     x.clo1 = this.clo1;
     *     x.l0 = this.l0;
     *     x.l1 = this.l1;
     *     x.l2 = this.l2;
     *     return x;
     *   }
-    *
-    *   // Only for closures
-    *   public Clo2$Obj$Int$Obj getUniqueThreadClosure() {
-    *     Clo$example$152 x = new Clo$example$152();
-    *     x.clo0 = this.clo0;
-    *     x.clo1 = this.clo1;
-    *     return x;
-    *   }
     * }
     * }}}
-    *
-    * For control-pure functions, the generated code deviates slightly, as the `applyFrame` method is replaced
-    * by a static method called `directApply` that accepts the concrete parameters of the Flix function.
-    * The `invoke` method then forwards its call to `directApply`, passing the values from the corresponding fields.
+    */
+  private def genControlImpureFunction(className: JvmName, defn: Def)(implicit root: Root, flix: Flix): Array[Byte] = {
+    val visitor = AsmOps.mkClassWriter()
+
+    // Header
+    val functionInterface = JvmOps.getErasedFunctionInterfaceType(defn.arrowType).jvmName
+    val frameInterface = BackendObjType.Frame
+    visitor.visit(AsmOps.JavaVersion, ACC_PUBLIC + ACC_FINAL, className.toInternalName, null,
+      functionInterface.toInternalName, Array(frameInterface.jvmName.toInternalName))
+
+    // Fields
+    for ((x, i) <- defn.lparams.zipWithIndex) {
+      visitor.visitField(ACC_PUBLIC, s"l$i", JvmOps.getErasedJvmType(x.tpe).toDescriptor, null, null)
```

Comment:
Sorry, I changed code under you. `JvmOps.getErasedJvmType(x.tpe).toDescriptor` is now `BackendType.toErasedBackendType(x.tpe).toDescriptor`. JvmType is gone

Reply:
Sneaky